<a href="https://colab.research.google.com/github/izzul112/ML_BMLT-Proyek_Akhir-Sistem_Rekomendasi-Anime/blob/main/Proyek_Akhir_Sistem_Rekomendasi.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Laporan Proyek Machine Learning - Habib Azizul Haq**

# Project Overview

Anime sebuah film animasi dari Jepang yang sangat populer dan mungkin banyak dari sebagian kalian yang merasa nostalgia ketika waktu kecil sering menontonnya di televisi. Sekarang untuk menontonnya banyak di antara kita yang sering mengunjungi situs-situs *streaming* seperti Bsite, IQiy, atau bahkan di kanal Youtube seperti Ani-One, Muse Indonesia dan Muse Asia.

Kepopuleran Anime saat memang lagi tinggi-tingginya, hal ini disebabkan banyak judul-judul besar Anime yang sudah punya basis penggemar yang besar, di musim panas 2022 ini banyak yang mendapat lanjutan musim keduanya bahkan Anime dengan judul Overlord mendapat kelanjutan musim keempat, dimana kelanjutan series ini sangat di tunggu-tunggu, perilisan musim keempat ini juga di iringi dengan rilisnya volume terbaru dari Light Novel Overlord yaitu volume 15 dan 16, tentu ini makin menambah *hype* dari judul Anime ini.


![image](https://user-images.githubusercontent.com/43197282/182755994-e4e082a1-d997-4b59-bffd-f675c4973c1b.png)
![image](https://user-images.githubusercontent.com/43197282/182756138-d9a75610-75c8-4c5a-a645-f3c4694ab197.png)
![image](https://user-images.githubusercontent.com/43197282/182756296-1afc570b-09c7-456c-a2ce-12fa73adf741.png)


Melihat tingginya kepopuleran yang terjadi sekarang juga membawa berkah tersendiri bagi industri *streaming* untuk medulang cuan yang banyak. Dengan adanya situs-situs streaming seperti disebutkan diatas membantu kita sebagai orang yang sedang mencari Anime untuk menonton secara **Legal**. Lantas ketika Anda mengunjungi katakanlah kanal youtube Muse Indonesia dan meng-klik salah-satu judul Anime disana, Anda akan melihat tampilan video lain yang serupa ketika Anda menggeser kebawah. Pernahkan Anda berfikir bagaimana video dibawah tersebut bisa muncul? Apa mereka muncul secara acak? *Tentu tidak demikian ferguso*.

Video dibawah muncul melalui serangkaian algoritma yang sering di sebut dengan sistem rekomendasi, ia merekomendasikan sebuah video yang mungkin menarik bagi Anda, supaya Anda meng-klik video tersebut. Lantas sekarang bayangkan Anda orang di bagian marketing dari salah-satu situs *streming* diatas, tentu Anda ingin meningkatkan pengalaman pengguna ketika menggunakan situs Anda, supaya pengguna merasa nyaman dan ingin berlangganan tetap di situs *streaming* Anda.

Sekarang Anda mengajukan sebuah proposal kepada atasan Anda untuk membuat sistem rekomendasi dan menerapkannya di situs *streaming*, Anda mengatakan alasan penting mengapa perlu sistem rekomendasi di situs *streaming* saat ini:

- Untuk meningkatkan kepusaan pengguna, sehingga pengguna betah berlama-lama mengunjungi situs *streaming* Anime perusahaan.

Dengan harapan bahwa hal itu bisa meningkatkan keuntungan bagi perusahaan dan tentu saja jika ini sukses di implementasikan dan menarik lebih banyak pelanggan maka tentu gaji Anda akan di naikkana oleh perusahhana Anda, karena kontribusi Anda.

- [Pentingnya Sistem Rekomendasi](https://fib.bunghatta.ac.id/index.php/en/article/209-pentingnya-sistem-informasi-di-era-digital#:~:text=Dengan%20menerapkan%20sebuah%20sistem%20informasi,dalam%20pasar%20yang%20selalu%20berubah.)

# Business Understanding

**Problem Statements**

Menjelaskan pernyataan masalah latar belakang:

*  Saat ini Anda belum memiliki model machine learning untuk memberikan rekomendasi Anime.

**Goals**

Menjelaskan tujuan dari pernyataan masalah:

*   Anda ingin menghasilkan model machine learning yang bisa memberikan rekomendasi Anime.

**Solution statements**

*   Kita akan menggunkana pendekatan Sistem Rekomendasi: Content-based Filtering, untuk memberikan rekomendasi Anime.
*   Kita akan menggunkana metrik `Precision` untuk menilai seberapa akurat hasil rekomendasi dari sistem rekomendasi yang kita buat.


## Data Understanding

Kita akan menggunkan dataset yang berisi 23 variabel yang sering muncul ketika Anda melihat deskripsi suatu Anime di situs MyAnimeList [1]. File dataset yang akan kita gunakan dalam proyek ini adalah anime_list.csv yang berisi 23 vaiabel dan 18.000 judul Anime.

Sumber dataset yang akan kita gunakan berasal dari [Kaggle](https://www.kaggle.com/datasets/snehaanbhawal/anime-list-for-recommendation-system-june-2021). Yang di publis oleh [SNEHAAN BHAWAL](https://www.kaggle.com/snehaanbhawal).

**Variabel-variabel pada *Anime List for Recommendation System* dataset adalah sebagai berikut:**

- mal_id : id unik dari situs myanimelist.net yang mengidentifikasi Anime.
- title : Judul dari Anime
- synopsis : Sinopsis, ringkasan singkat yang memberikan gambaran tentang suatu karya.
- background : Latar belakang.
- aired : Rentang waktu anime ditayangkan. contoh format: "30 Sep 2004 sampai 29 Sep 2005".
- airing : Apakah saat ini sedang ditayangkan atau tidak. 0 jika anime telah berakhir 1 jika sedang berlangsung
- duration : Durasi per-episode.
- episodes : Jumlah episode dari suatu judul Anime.
- type : Jenis Anime seperti: ['TV', 'Movie', 'OVA', 'Special', 'ONA', 'Music', 'Unknown']
- favorites : Jumlah orang yang memfavoritkan anime ini.
- members : Jumlah anggota.
- rank : Peringkat per Juni 2021.
- popularity : Peringkat Popularitas per Juni 2021.
- score : Skor rata-rata yang diberikan oleh pengguna.
- scored_by : Jumlah orang yang merating Anime.
- rating : Peringkat diberikan untuk anime.['PG - Anak', 'PG-13 - Remaja 13 tahun ke atas', 'G - Semua Usia', 'R+ - Ketelanjangan Ringan', 'R - 17+','Rx - Hentai' ]
- premiered : Tahun dan musim pemutaran perdana.
- genres : Genre Anime.
- related : Anime terkait.
- status : Status Anime apakah sedang tayang saat ini atau tidak.
- licensors : Nama Pemberi Lisensi.
- producers : Nama Produser.
- studios : Nama studio yang menggerjakan.

Kita akan melakukan sedikit analisis untuk mengethui beberapa hal tentang variabel diatas diantaranya :

1. Kita akan melihat ada berapa jumlah baris dari data dalam tampilan tabel.
2. Kita akan melihat tipe data dari 21 kolom di dataset.
3. Kita akan mengecek deskripsi statistik data.
4. Kita akan melihat apakah ada data yang kosong / tidak ada isinya.

Sebelum kita mulai mengerjakan kita import semua library yang di butuhkan.



In [1]:
# Import library
import pandas as pd
import numpy as np 
import tensorflow as tf
from tensorflow import keras
from keras import layers
from pathlib import Path
import matplotlib.pyplot as plt

Kita sambungkan dulu Google Colab kita dengan Google Drive, karena dataset kita berada di Google Drive.

In [2]:
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


Setelah itu kita buat variabel untuk menampung dataset dengan nama data_anime untuk menampung data anime_list.csv yang akan kita panggil dari Google Drive yang sudah kita sambungkan dengan Google Colab.



In [3]:
data_anime = pd.read_csv('/content/gdrive/My Drive/dataset/anime_list.csv')

# **Pembahasan**
**1. Kita akan melihat ada berapa jumlah baris dari data dalam tampilan tabel.**

Kita akan menggunakan menggunakan kode `data_anime`, `data_anime` adalah nama variabel yang kita ganakan saat meload dataset anime_list.csv.

In [4]:
data_anime

mal_id                                              title  \
0           1                                       Cowboy Bebop   
1           5                    Cowboy Bebop: Tengoku no Tobira   
2           6                                             Trigun   
3           7                                 Witch Hunter Robin   
4           8                                     Bouken Ou Beet   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
0      In the year 2071, humanity has colonized sever...          -   
1      Another day, another bounty—such is the life o...          -   
2      Vash the Stampede is the man with a $$60,000,0...          -   
3      Witches are individuals with special powers li...          -   
4      It is the dark century and the people are suff...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
0       Apr 3, 1998 to Apr 24, 1999       0  24 min per ep        26       TV   
1                       Sep 1, 2001       0    1 hr 55 min         1    Movie   
2       Apr 1, 1998 to Sep 30, 1998       0  24 min per ep        26       TV   
3       Jul 2, 2002 to Dec 24, 2002       0  25 min per ep        26       TV   
4      Sep 30, 2004 to Sep 29, 2005       0  23 min per ep        52       TV   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                          rating  \
0          59968  ...   8.77     661519  R - 17+ (violence & profanity)   
1           1063  ...   8.39     168515  R - 17+ (violence & profanity)   
2          11882  ...   8.23     288760       PG-13 - Teens 13 or older   
3            512  ...   7.27      37135       PG-13 - Teens 13 or older   
4             10  ...   6.97       5463                   PG - Children   
...          ...  ...    ...        ...                             ...   
18157          0  ...  -1.00         -1       PG-13 - Teens 13 or older   
18158          0  ...  -1.00         -1                    G - All Ages   
18159          0  ...  -1.00         -1                            None   
18160          2  ...  -1.00         -1       PG-13 - Teens 13 or older   
18161          0  ...  -1.00         -1                R+ - Mild Nudity   

         premiered                                             genres  \
0      Spring 1998    Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                -              Action, Drama, Mystery, Sci-Fi, Space   
2      Spring 1998  Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Summer 2002  Action, Mystery, Police, Supernatural, Drama, ...   
4        Fall 2004          Adventure, Fantasy, Shounen, Supernatural   
...            ...       

Dari kode `data_anime` kita mendapat informasi:

*   Ada 18.162 baris dataset anime_list.csv kita
*   Dan terdapat 23 kolom



**2. Kita akan melihat tipe data dari kedua dataset kita.**

Kita akan menggunakan `data_anime.info()`, `.info()` akan menampilkan informasi tipe data 7 kolom dataset anime.csv yang kita gunakan.

In [5]:
data_anime.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18162 entries, 0 to 18161
Data columns (total 23 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   mal_id      18162 non-null  int64  
 1   title       18162 non-null  object 
 2   synopsis    18162 non-null  object 
 3   background  18162 non-null  object 
 4   aired       18162 non-null  object 
 5   airing      18162 non-null  int64  
 6   duration    18162 non-null  object 
 7   episodes    18162 non-null  int64  
 8   type        18162 non-null  object 
 9   favorites   18162 non-null  int64  
 10  members     18162 non-null  int64  
 11  rank        18162 non-null  float64
 12  popularity  18162 non-null  int64  
 13  score       18162 non-null  float64
 14  scored_by   18162 non-null  int64  
 15  rating      18162 non-null  object 
 16  premiered   18162 non-null  object 
 17  genres      18095 non-null  object 
 18  related     18162 non-null  object 
 19  status      18162 non-nul

Dari *output* `data_anime.info()`, terlihat bahwa:

*   Terdapat 2 kolom dengan tipe data float64, yaitu: 
*   Terdapat 7 kolom dengan tipe data int64, yaitu: 
*   Terdapat 14 kolom dengan tipe data object, yaitu: 



**3. Kita akan mengecek deskripsi statistik data.**

Kita akan mengeceknya dengan fungsi `describe()`.

In [6]:
data_anime.describe()

mal_id        airing      episodes      favorites       members  \
count  18162.000000  18162.000000  18162.000000   18162.000000  1.816200e+04   
mean   22321.114470      0.017564     11.945436     424.283889  3.698543e+04   
std    15330.497714      0.131364     48.074396    3858.020730  1.372226e+05   
min        1.000000      0.000000     -1.000000       0.000000  1.100000e+01   
25%     6297.500000      0.000000      1.000000       0.000000  3.330000e+02   
50%    24086.000000      0.000000      2.000000       3.000000  2.012500e+03   
75%    36431.750000      0.000000     12.000000      27.000000  1.338775e+04   
max    48960.000000      1.000000   3057.000000  175658.000000  2.884844e+06   

               rank    popularity         score     scored_by  
count  18162.000000  18162.000000  18162.000000  1.816200e+04  
mean    7418.114029   9091.728554      4.301493  1.833681e+04  
std     5125.230109   5252.673268      3.483474  7.974844e+04  
min       -1.000000      1.000000     -1.000000 -1.000000e+00  
25%     2782.500000   4544.250000     -1.000000 -1.000000e+00  
50%     7328.000000   9087.000000      6.020000  5.850000e+02  
75%    11868.750000  13637.750000      6.830000  4.714000e+03  
max    16464.000000  18247.000000      9.170000  1.983096e+06

Fungsi `describe()` memberikan informasi statistik pada masing-masing kolom, antara lain:

*   Count  adalah jumlah sampel pada data.
*   Mean adalah nilai rata-rata.
*   Std adalah standar deviasi.
*   Min yaitu nilai minimum setiap kolom.
*   25% adalah kuartil pertama. Kuartil adalah nilai yang menandai batas interval dalam empat bagian sebaran yang sama.
*   50% adalah kuartil kedua, atau biasa juga disebut median (nilai tengah).
*   75% adalah kuartil ketiga.
*   Max adalah nilai maksimum.

Perhatikan pada kolom `score` bisa kita lihat bahwa baris Max atau nilai maksimum yang diberikan oleh rata-rata user adalah 9.17 dan baris Min yaitu nilai terendahnya adalah -1. Niali -1 terjadi ketika user melihat Anime tersebut tapi tidak memberikan rating.

**4. Kita akan melihat apakah ada data yang kosong / tidak ada isinya.**

Kita akan mengeceknya dengan fungsi `.isnull().sum()`. Fungsi `.isnull()` akan mengecek apakah ada data kosng pada setiap baris pada semua kolom di dataset kita, kemudian kita gunakan juga fungsi `.sum()` untuk menjumlahnya sehingga hasilnya akan seperti berikut.

In [7]:
data_anime.isnull().sum()

mal_id         0
title          0
synopsis       0
background     0
aired          0
airing         0
duration       0
episodes       0
type           0
favorites      0
members        0
rank           0
popularity     0
score          0
scored_by      0
rating         0
premiered      0
genres        67
related        0
status         0
licensors      0
producers      0
studios        0
dtype: int64

Dari hasil di atas kita tau bahwa di kolom genres memiliki 67 data kosong.

# **Data Preparation**

Setelah kita melalui proses Data Understanding, kita menjadi lebih paham tentang seperti apa isi dataset kita.

Oleh karena itu di proses ini kita akan melakukan 2 proses yaitu:

1.  Kita akan mengubah nilai score yang -1 menjadi kosong supaya bisa kita hapus.
2.  Kita akan menghapus data yang kosong.

**1.  Kita akan mengubah nilai score yang -1 menjadi kosong supaya bisa kita hapus.**

Penghapusan ini penting mengingat nilai -1 terjadi karena user menonton anime tersebut tapi tidak memberikan score, maka dari itu supaya nilai -1 ini tidak mempengaruhi sitem rekomendasi kita, kita akan menggantinya dengan nilai NaN / kosong supaya bisa kita hapus baris data yang mengandung nilai -1.

Sebelum itu kita lihat dulu nilai ada berapa jumlah nilai -1 nya.

Kita akan melihatnya dengan kode `data_anime.loc[(data_anime['score']==-1)]`.
- `.loc[]` adalah akses grub baris dan kolom meurut label, dimana kita ingin mengakses kolom score dari data_anime.
- `==-1` adalah parameter untuk hanya menampilkan kolom score dengan nilai yang sama dengan -1.

In [8]:
data_anime.loc[(data_anime['score']==-1)]

mal_id                                              title  \
1574     1739                         Shibawanko no Wa no Kokoro   
1695     1863                            Silk Road Shounen Yuuto   
1986     2173  Kouchuu Ouja Mushiking Super Battle Movie: Yam...   
2161     2366                                Touma Kishinden Oni   
2418     2636          Dorami-chan: Wow, The Kid Gang of Bandits   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
1574   Based on a japanese children`s book by Yoshie ...          -   
1695   When a boy Yuto visits Qinghai in China, he is...          -   
1986   A shadow of Darkness approaches a peaceful For...          -   
2161   Shuramaru is hated and feared by the villagers...          -   
2418   No synopsis has been added for this series yet...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
1574    Apr 5, 2006 to Mar 14, 2007       0   2 min per ep        80       TV   
1695   Sep 16, 2006 to Mar 24, 2007       0  20 min per ep        26       TV   
1986                   Mar 21, 2007       0         50 min         1    Movie   
2161    Oct 5, 1995 to Mar 21, 1996       0   8 min per ep        25       TV   
2418                    Mar 9, 1991       0         40 min         1    Movie   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                     rating  \
1574           0  ...   -1.0         -1               G - All Ages   
1695           0  ...   -1.0         -1                       None   
1986           1  ...   -1.0         -1               G - All Ages   
2161           0  ...   -1.0         -1  PG-13 - Teens 13 or older   
2418           0  ...   -1.0         -1               G - All Ages   
...          ...  ...    ...        ...                        ...   
18157          0  ...   -1.0         -1  PG-13 - Teens 13 or older   
18158          0  ...   -1.0         -1               G - All Ages   
18159          0  ...   -1.0         -1                       None   
18160          2  ...   -1.0         -1  PG-13 - Teens 13 or older   
18161          0  ...   -1.0         -1           R+ - Mild Nudity   

         premiered                                             genres  \
1574   Spring 2006                                               Kids   
1695     Fall 2006                     Adventure, Fantasy, Historical   
1986             -                                 Adventure, Fantasy   
2161     Fall 1995                                       Supernatural   
2418             -                                            Fantasy   
...            ...                                                ...   
18157       

Dari hasil kode diatas kita bisa lihat bahwa:
- Ada 5.296 baris yang mengandung nilai -1 di kolom score.

Lanjut keproses mengubah nilai -1 menjadi Nan.

- Pertama kita buat variabel baru dengan nama `anime_clean` yang akan diisi dengan data_anime dengan kode `anime_clean = data_anime.copy()`.

- Setelah itu kita mengganti nilai -1 dengan fungsi `.replace({-1: np.nan}, inplace=True)`, nilai di dalam `{-1: np.nan}` berarti -1 diganti dengan nilai NaN dan `inplace=True` supaya jika kita berhasil mengganti nilai rating_user mengembalinkan nilai None / tidak ada.

- Setelah itu kita panggil variabel `data_anime_clean` untuk melihat apakah sudah berubah?

In [9]:
data_anime_clean = data_anime.copy()
data_anime_clean["score"].replace({-1: np.nan}, inplace=True)
data_anime_clean

mal_id                                              title  \
0           1                                       Cowboy Bebop   
1           5                    Cowboy Bebop: Tengoku no Tobira   
2           6                                             Trigun   
3           7                                 Witch Hunter Robin   
4           8                                     Bouken Ou Beet   
...       ...                                                ...   
18157   48952              Twisted Wonderland 1st Anniversary PV   
18158   48953  Natsume Yuujinchou: Hitoyoshi Kuma de no Yasas...   
18159   48954                                     Hairpin Double   
18160   48956                        Wu Shang Shen Di 2nd Season   
18161   48960                                             Hunter   

                                                synopsis background  \
0      In the year 2071, humanity has colonized sever...          -   
1      Another day, another bounty—such is the life o...          -   
2      Vash the Stampede is the man with a $$60,000,0...          -   
3      Witches are individuals with special powers li...          -   
4      It is the dark century and the people are suff...          -   
...                                                  ...        ...   
18157  Disney's Twisted Wonderland mobile game had it...          -   
18158  A collaboration with the Kumamoto prefecture i...          -   
18159                                                  -          -   
18160                                                  -          -   
18161  Hunter is looking for Gossip every night. A sh...          -   

                              aired  airing       duration  episodes     type  \
0       Apr 3, 1998 to Apr 24, 1999       0  24 min per ep        26       TV   
1                       Sep 1, 2001       0    1 hr 55 min         1    Movie   
2       Apr 1, 1998 to Sep 30, 1998       0  24 min per ep        26       TV   
3       Jul 2, 2002 to Dec 24, 2002       0  25 min per ep        26       TV   
4      Sep 30, 2004 to Sep 29, 2005       0  23 min per ep        52       TV   
...                             ...     ...            ...       ...      ...   
18157                  Mar 17, 2021       0          2 min         1  Special   
18158                  Apr 21, 2021       0          2 min         1      ONA   
18159                     2021 to ?       0        Unknown        -1       TV   
18160             May 21, 2021 to ?       1   7 min per ep        64      ONA   
18161                  Sep 14, 2017       0          8 min         1    Movie   

       favorites  ...  score  scored_by                          rating  \
0          59968  ...   8.77     661519  R - 17+ (violence & profanity)   
1           1063  ...   8.39     168515  R - 17+ (violence & profanity)   
2          11882  ...   8.23     288760       PG-13 - Teens 13 or older   
3            512  ...   7.27      37135       PG-13 - Teens 13 or older   
4             10  ...   6.97       5463                   PG - Children   
...          ...  ...    ...        ...                             ...   
18157          0  ...    NaN         -1       PG-13 - Teens 13 or older   
18158          0  ...    NaN         -1                    G - All Ages   
18159          0  ...    NaN         -1                            None   
18160          2  ...    NaN         -1       PG-13 - Teens 13 or older   
18161          0  ...    NaN         -1                R+ - Mild Nudity   

         premiered                                             genres  \
0      Spring 1998    Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                -              Action, Drama, Mystery, Sci-Fi, Space   
2      Spring 1998  Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Summer 2002  Action, Mystery, Police, Supernatural, Drama, ...   
4        Fall 2004          Adventure, Fantasy, Shounen, Supernatural   
...            ...       

Hasil di atas menunjukkan bahwa kita berhasil merubah nilai yang tadinya -1 menjadi NaN atau kosong. Sebelum lanjut kita cek lagi apakah nilai -1 nya masih ada atau tidak dengan kode `data_anime_clean.loc[(data_anime_clean['score']==-1)]`



In [10]:
data_anime_clean.loc[(data_anime_clean['score']==-1)]

Empty DataFrame
Columns: [mal_id, title, synopsis, background, aired, airing, duration, episodes, type, favorites, members, rank, popularity, score, scored_by, rating, premiered, genres, related, status, licensors, producers, studios]
Index: []

[0 rows x 23 columns]

Dari hasil diatas kita bisa yakin bahwa sudah tidak ada nilai -1 di kolom score. Selanjutnya kita akan melanjutkan ke proses selanjutnya.

**2.  Kita akan menghapus data yang kosong.**

Pertama kita cek lagi data yang kosong kita dengan kode `data_anime_clean.isnull().sum()`.

In [11]:
data_anime_clean.isnull().sum()

mal_id           0
title            0
synopsis         0
background       0
aired            0
airing           0
duration         0
episodes         0
type             0
favorites        0
members          0
rank             0
popularity       0
score         5296
scored_by        0
rating           0
premiered        0
genres          67
related          0
status           0
licensors        0
producers        0
studios          0
dtype: int64

Dari hasil output diatas kita jadi tau bahwa ada tambahan data kosng yaitu pada kolom score sebanyak 5296.

Setelah berhasil kita hapus data kosong dengan kode `data_anime_clean = data_anime_clean.dropna(axis = 0, how ='any')`.

- Dimana `axis = 0` berarti hanya mengapus baris yang mengandung nilai kosong.
- Sedang `how = 'any'` berarti akan menghapus juga jika ada baris atau kolom yang nilainya kosong semua.

In [12]:
data_anime_clean = data_anime_clean.dropna(axis = 0, how ='any')

Setelah kita jalankan kode diatas kita cek lagi apakah datanya yang kosong sudah terhapus? Kita cek dengan kode `data_anime_clean.isnull().sum()`.

In [13]:
data_anime_clean.isnull().sum()

mal_id        0
title         0
synopsis      0
background    0
aired         0
airing        0
duration      0
episodes      0
type          0
favorites     0
members       0
rank          0
popularity    0
score         0
scored_by     0
rating        0
premiered     0
genres        0
related       0
status        0
licensors     0
producers     0
studios       0
dtype: int64

Hasilnya sudah tidak ada lagi data kosong di variabel data_anime_clean.

Sebelum lanjut ke proses Modeling, kita akan menghapus beberapa fitur di variabel data_anime_clean.

Kita devinisikan variabel baru dulu dengan nama data_anime_cleaned yang akan menampung data dari data_anime_clean yang sudah di hapus beberapa kolomnya.

Kita akan menghapus kolom:
- duration.
- airing.
- aired.
- background.
- premiered.
- licensors.
- producers.
- status.

Setelah itu kita panggi lagi data_anime_cleaned untuk mengetahui apa isinya.

In [14]:
data_anime_cleaned = data_anime_clean.copy().drop(columns = ['duration','airing','aired','background','premiered','licensors','producers','status'], axis = 1)
data_anime_cleaned

mal_id                            title  \
0           1                     Cowboy Bebop   
1           5  Cowboy Bebop: Tengoku no Tobira   
2           6                           Trigun   
3           7               Witch Hunter Robin   
4           8                   Bouken Ou Beet   
...       ...                              ...   
18052   48653                   Yoru Ni Kakeru   
18053   48654              Bitch na Inane-sama   
18071   48727                     Yoru no Kuni   
18072   48735                   Chibi Revenger   
18126   48876                   Yoru wa Honoka   

                                                synopsis  episodes   type  \
0      In the year 2071, humanity has colonized sever...        26     TV   
1      Another day, another bounty—such is the life o...         1  Movie   
2      Vash the Stampede is the man with a $$60,000,0...        26     TV   
3      Witches are individuals with special powers li...        26     TV   
4      It is the dark century and the people are suff...        52     TV   
...                                                  ...       ...    ...   
18052  Music video of the song "Yoru Ni Kakeru" by Yo...         1  Music   
18053                                                  -        -1    OVA   
18071  Singer Aimer and Director Ryo-timo's original ...        -1    ONA   
18072  Mini anime for Tokyo Revengers released on ani...        -1    ONA   
18126  Music video for the song Yoru wa Honoka by Eve...         1  Music   

       favorites  members    rank  popularity  score  scored_by  \
0          59968  1366272    32.0          44   8.77     661519   
1           1063   291258   163.0         542   8.39     168515   
2          11882   591134   285.0         213   8.23     288760   
3            512    97700  2508.0        1535   7.27      37135   
4             10    13576  3769.0        4514   6.97       5463   
...          ...      ...     ...         ...    ...        ...   
18052         21     1706   465.0        9563   8.06       1094   
18053          3     1530    -1.0        9819   5.93        508   
18071         19     3783  1920.0        7517   7.40        810   
18072          4     1413  4773.0        9997   6.74        313   
18126         17      586  1348.0       12109   7.58        363   

                               rating  \
0      R - 17+ (violence & profanity)   
1      R - 17+ (violence & profanity)   
2           PG-13 - Teens 13 or older   
3           PG-13 - Teens 13 or older   
4                       PG - Children   
...                               ...   
18052       PG-13 - Teens 13 or older   
18053                     Rx - Hentai   
18071                    G - All Ages   
18072       PG-13 - Teens 13 or older   
18126       PG-13 - Teens 13 or older   

                                                  genres  \
0        Action, Adventure, Comedy, Drama, Sci-Fi, Space   
1                  Action, Drama, Mystery, Sci-Fi, Space   
2      Action, Sci-Fi, Adventure, Comedy, Drama, Shounen   
3      Action, Mystery, Police, Supernatural, Drama, ...   
4              Adventure, Fantasy, Shounen, Supernatural   
...                                                  ...   
18052                                              Music   
18053                                             Hentai   
18071                                            Fantasy   
18072                                             Comedy   
18126                                Music, Supernatural   

                                                 related         studios  
0      {'Adaptation': [{'mal_id': 173, 'type': 'manga...         Sunrise  
1      {'Parent story': [{'mal_id': 1, 'type': 'anime...           Bones  
2      {'Adaptation': [{'mal_id': 703, 'type': 'manga...        Madhouse  
3                                                     {}         Sunrise  
4      {'Adaptation': [{'mal_id': 1348, 'type': 'mang...  Toei Animation  
...         

Seperti yang kita lihat kolom yang tadi kita hapus sudah tidak ada, hanya menyisakan 12847 baris dan 15 kolom.

# **Modeling**

Pada tahap modeling kita akan menggunakan pendekatan Content-based Filtering, pendekatan ini bekerja dengan mencari kemiripan dari setiap Anime yang dilihat oleh pengguna, memiliki beberapa kelebihan diantaranya:

- Pengguna tau bahwa ada Anime dengan fitur yang sama namun dengan judul yang berbeda.

Jika ada kelebihan tentu ada kekurangan ya, kekurangan dari pendekatan ini adalah:

- Hanya berfokus pada kemiripan kata kunci.

Baik, untuk menerapkan pendekatan Content-based Filtering kita akan melakuka:

1. Kita akan mencari representasi fitur synopsis menggunkana `TfidfVectorizer`

2. Mencari pasangan rekomendasi menggunkan metrik pairwise yang berfungsi untuk representasi koleksi vektor menggunakan `sigmoid_kernel`.

**Langsung saja kita mulai dari yang pertama yaitu:**

**1. Kita akan mencari representasi fitur synopsis menggunkana `TfidfVectorizer`**
 
Pertama kita import dulu `TfidfVectorizer` kemudian masukkan ke variabel tfv,
setelah itu kita set variabel dari `TfidfVectorizer`-nya dengan keterangan seperti ini:
- `min_df=5` : untuk mengabaikan kata/istilah yang jarang muncul.
- `max_feature=None` : supaya tidak membuat kosakata yang dengan berdasar frekuensi kemunculan teratas.
- `strip_accent='unicode'` : untuk menormalisasi karakter, kita memilih 'unicode' karena berkerja pada karakter apapun.
- `analyzer='word'` : untuk menggunkan ngram kita set analyzer ke word.
- `token_pattern=r'\w{1,}'` : Ekspresi reguler yang menunjukkan apa yang merupakan sebuah "token", hanya digunakan jika analyzer == 'word'.
- `ngram_range=(1, 3)` : Batas bawah dan batas atas kisaran nilai-n untuk n-gram berbeda yang akan diekstraksi.
- `stop_word='english'` : untuk mendeteksi kata akhiran dalam bahasa Inggris.

Setelah itu kita buat variabel dengan nama synopsis_str yang akan menampung fitur synopsis dari variabel anime_cleaned.

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer


tfv = TfidfVectorizer(min_df=5, max_features=None,
            strip_accents='unicode', analyzer='word',token_pattern=r'\w{1,}',
            ngram_range=(1, 3),
            stop_words = 'english')

synopsis_str = data_anime_cleaned['synopsis']

Selanjutnya kita lakukan proses fit, untuk mempelajari kosa kata dan Inverse Document Frequency.

Setelah itu kita lihat ukurannya dengan fungsi `.shape`.

In [16]:
tfv_matrix = tfv.fit_transform(synopsis_str)
tfv_matrix.shape

(12847, 17792)


Dan hasilnya seperti diatas.

**2. Mencari pasangan rekomendasi menggunkan metrik pairwise yang berfungsi untuk representasi koleksi vektor menggunakan `sigmoid_kernel`.**

Kita import dulu metriknya, kita buat variabel sig untuk menampung hasil keluaran dari `sigmoid_kernel`, setelah itu kita masukkan variabel tfv_matrik yg berisi hasil dari proses fit_transform.

In [17]:
from sklearn.metrics.pairwise import sigmoid_kernel

# Compute the sigmoid kernel
sig = sigmoid_kernel(tfv_matrix, tfv_matrix)

Setelah itu kita panggil untuk melihat seperti apa bentuknya.

In [18]:
sig

array([[0.76161776, 0.76159987, 0.76159496, ..., 0.76159495, 0.76159421,
        0.76159442],
       [0.76159987, 0.76161776, 0.76159564, ..., 0.76159416, 0.76159416,
        0.76159416],
       [0.76159496, 0.76159564, 0.76161776, ..., 0.7615949 , 0.76159427,
        0.76159416],
       ...,
       [0.76159495, 0.76159416, 0.7615949 , ..., 0.76161776, 0.76159495,
        0.76159622],
       [0.76159421, 0.76159416, 0.76159427, ..., 0.76159495, 0.76161776,
        0.76159416],
       [0.76159442, 0.76159416, 0.76159416, ..., 0.76159622, 0.76159416,
        0.76161776]])

`sigmoid_kernel` menghasilkan keluaran berupa array multi dimensi yang akan kita cocokan nanti saat proses memberikan rekomendasi.

Sekarang karena kita buat variabel baru bernama data_anime_index untuk menerapkan index pada 'title'.

In [19]:
data_index_anime = pd.Series(data_anime_cleaned.index, index=data_anime_cleaned['title'])

Dan selanjutnya kita panggil variabel data_anime_index.

In [20]:
data_index_anime

title
Cowboy Bebop                           0
Cowboy Bebop: Tengoku no Tobira        1
Trigun                                 2
Witch Hunter Robin                     3
Bouken Ou Beet                         4
                                   ...  
Yoru Ni Kakeru                     18052
Bitch na Inane-sama                18053
Yoru no Kuni                       18071
Chibi Revenger                     18072
Yoru wa Honoka                     18126
Length: 12847, dtype: int64

Seperti yang kita lihat di atas angka di sebelah kanan adalah index dari title anime.

Lanjut kita buat fungsi yang akan memberikan rekomendasi dengan paramenter title.

In [21]:
def give_recomendation(title, sig = sig):
    # Dapatkan indeks yang sesuai dengan title asli
    idx = data_index_anime[title]

    # Dapatkan skor kesamaan pairwsie
    sig_scores = list(enumerate(sig[idx]))

    # Urutkan keluaran film
    sig_scores = sorted(sig_scores, key=lambda x: x[1], reverse=True)

    # Pilih dari 5 Anime paling mirip
    sig_scores = sig_scores[1:6]

    # Mengeluarkan indeks Anime
    anime_index = [i[0] for i in sig_scores]

    # Menampilkan 5 anime paling mirip
    return pd.DataFrame({'Anime name': data_anime_cleaned['title'].iloc[anime_index].values,
                                 'Rating': data_anime_cleaned['score'].iloc[anime_index].values})

In [22]:
give_recomendation('Cowboy Bebop')

Anime name  Rating
0       Cowboy Bebop: Tengoku no Tobira    8.39
1                  Ginga Senpuu Braiger    6.61
2      Cowboy Bebop: Ein no Natsuyasumi    6.26
3             Seihou Bukyou Outlaw Star    7.86
4  Mutant Turtles: Choujin Densetsu-hen    5.26

In [23]:
give_recomendation('Mobile Suit Gundam')

Anime name  Rating
0                        Mobile Suit Gundam I    7.29
1                     Mobile Suit Zeta Gundam    7.92
2                      Mobile Suit Gundam G40    6.13
3    Mobile Suit Gundam 0083: Stardust Memory    7.29
4  Mobile Suit Gundam 0080: War in the Pocket    7.91

# **Evaluation**

Ditahap kita akan mengevaluasi model sistem rekomendasi kita dengan metrik `Precision`.

Pertama kita coba panggil fungsi rekomendasinya seperti kode di bawah, untuk menghasilkan rekomendasi dan setelah itu kita hitung menggunkaan metrik `Precision`.

Dari dua hasil keluaran diatas kita akan menghitungnya dengan rumus Precision = jumlah Anime yang relevan dibagi jumlah Anime yang direkomendasikan.

1. Hasil rekomendasi untuk 'Coboy Bebop' adalah:
    
    Precission = 4/5.

    Jadi presisinya = 80%

2. Hasil rekomendasi untuk 'Mobile Suit Gundam' adalah:

    Precission = 5/5.

    Jadi presisinya = 100%

# **Conclusion**

Akhirnya kita berhasil membuat sistem rekomendasi Anime dan mengujinya dengan hasil:
1. 80%.
2. 100%.

Jika kita rata-rata maka kita mendapat hasil akurasi 90%.

[enumerate()](https://realpython.com/python-enumerate/#using-pythons-enumerate)